session_db/pg_session_store.py

```diff
@@ -0,0 +1,169 @@
+# Copyright (c) Odoo SA 2017
+# @author Nicolas Seinlet
+# Copyright (c) ACSONE SA 2022
+# @author Stéphane Bidoul
+import json
+import logging
+import os
+
+import psycopg2
+
+import odoo
+from odoo import http
+from odoo.tools._vendor import sessions
+from odoo.tools.func import lazy_property
+
+_logger = logging.getLogger(__name__)
+
+lock = None
+if odoo.evented:
+    import gevent.lock
+
+    lock = gevent.lock.RLock()
+elif odoo.tools.config["workers"] == 0:
+    import threading
+
+    lock = threading.RLock()
+
+
+def with_lock(func):
+    def wrapper(*args, **kwargs):
+        try:
+            if lock is not None:
+                lock.acquire()
+            return func(*args, **kwargs)
+        finally:
+            if lock is not None:
+                lock.release()
+
+    return wrapper
+
+
+def with_cursor(func):
+    def wrapper(self, *args, **kwargs):
+        tries = 0
+        while True:
+            tries += 1
+            try:
+                self._ensure_connection()
+                return func(self, *args, **kwargs)
+            except (psycopg2.InterfaceError, psycopg2.OperationalError):
+                self._close_connection()
+                if tries > 4:
+                    _logger.warning(
+                        "session_db operation try %s/5 failed, aborting", tries
+                    )
+                    raise
+                _logger.info("session_db operation try %s/5 failed, retrying", tries)
+
+    return wrapper
+
+
+class PGSessionStore(sessions.SessionStore):
+    def __init__(self, uri, session_class=None):
+        super().__init__(session_class)
+        self._uri = uri
+        self._cr = None
+        self._open_connection()
+        self._setup_db()
+
+    def __del__(self):
+        self._close_connection()
+
+    @with_lock
+    def _ensure_connection(self):
+        if self._cr is None:
+            self._open_connection()
+
+    @with_lock
+    def _open_connection(self):
+        self._close_connection()
+        cnx = odoo.sql_db.db_connect(self._uri, allow_uri=True)
+        self._cr = cnx.cursor()
+        self._cr._cnx.autocommit = True
+
+    @with_lock
+    def _close_connection(self):
+        """Return cursor to the pool."""
+        if self._cr is not None:
+            try:
+                self._cr.close()
+            except Exception:  # pylint: disable=except-pass
+                pass
+            self._cr = None
+
+    @with_lock
+    @with_cursor
+    def _setup_db(self):
+        self._cr.execute(
+            """
+                CREATE TABLE IF NOT EXISTS http_sessions (
+                    sid varchar PRIMARY KEY,
+                    write_date timestamp without time zone NOT NULL,
+                    payload text NOT NULL
```

Comment:
We could use a jsonb column here nowadays

Reply:
What benefit do you expect of a Jsonb column in this case?

Reply:
Given payload is a JSON, it simply seems right to use the column type made for this

Reply:
I don't know. Intuitively I'd say a jsonb column would be less performant as postgres would need to parse it somehow. And we don't need to query the content of the sessions so that does not seem useful to me at first sight.